kedro/framework/cli/catalog.py

```diff
@@ -174,3 +176,56 @@ def _add_missing_datasets_to_catalog(missing_ds, catalog_path):
     catalog_path.parent.mkdir(exist_ok=True)
     with catalog_path.open(mode="w") as catalog_file:
         yaml.safe_dump(catalog_config, catalog_file, default_flow_style=False)
+
+
+@catalog.command("show")
+@env_option
+@click.pass_obj
+def show_catalog_datasets(metadata: ProjectMetadata, env):
+    session = _create_session(metadata.package_name, env=env)
+    context = session.load_context()
+    catalog_conf = context.config_loader["catalog"]
+    secho(yaml.dump(catalog_conf))
+
+
+@catalog.command("resolve")
+@env_option
+@click.pass_obj
+def resolve_catalog_datasets(metadata: ProjectMetadata, env):
+    session = _create_session(metadata.package_name, env=env)
+    context = session.load_context()
+    catalog_conf = context.config_loader["catalog"]
+
+    # Create a list of all datasets used in the project pipelines.
+    pipeline_datasets = []
+    for pipe in pipelines.keys():
+        pl_obj = pipelines.get(pipe)
+        if pl_obj:
+            pipeline_ds = pl_obj.data_sets()
+            for ds in pipeline_ds:
+                pipeline_datasets.append(ds)
+        else:
+            existing_pls = ", ".join(sorted(pipelines.keys()))
+            raise KedroCliError(
+                f"'{pipe}' pipeline not found! Existing pipelines: {existing_pls}"
+            )
+
+    # Create a copy of the catalog config to not modify the original.
+    catalog_copy = copy.deepcopy(catalog_conf)
+    # Loop over all entries in the catalog, find the ones that contain a pattern to be matched,
+    # loop over al datasets in the pipeline and match these against the patterns.
+    # Then expand the matches and add them to the catalog copy to display on the CLI.
+    for ds_name, ds_config in catalog_conf.items():
+        if "}" in ds_name:
+            for pipeline_dataset in set(pipeline_datasets):
+                result = parse(ds_name, pipeline_dataset)
+                if result:
+                    config_copy = copy.deepcopy(ds_config)
+                    # Match results to patterns in catalog entry
+                    for key, value in config_copy.items():
+                        if isinstance(value, Iterable) and "}" in value:
+                            string_value = str(value)
+                            config_copy[key] = string_value.format_map(result.named)
+                    catalog_copy[pipeline_dataset] = config_copy
+
```

Comment:
I find this extremely hard to read if not for the comments. I think it would be much better if the code actually does what the comment says, step by step, rather than nesting levels.

Moreover, this complexity seems to have hidden a logical incorrectness of the code: all datasets that appear exactly as they are in the catalog, shouldn't be matched against the patterns (which the original code doesn't do).

```suggestion
    skip = [ds_name for catalog_conf.keys() if "}" not in ds_name]
    datasets = set(pipeline_datasets) - set(skip)
    patterns = [ds_name for catalog_conf.keys() if "}" in ds_name]
    matches = [(ds, pattern, parse(pattern, ds)) for ds in datasets for pattern in patterns]
    matches = [(ds, pattern, result) for ds, pattern, result in matches if result]
    for ds, pattern, result in matches:
        cfg = copy.deepcopy(catalog_conf[pattern])
        # Match results to patterns in catalog entry
        for key, value in cfg.items():
            if isinstance(value, Iterable) and "}" in value:
                string_value = str(value)
                cfg[key] = string_value.format_map(result.named)
        catalog_copy[pipeline_dataset] = cfg

```

The code for populating the template should probably be a separate function for readability and also to ensure that we can call it recursively (as each value can be a dictionary with other values, etc).

The original code does all the matching and it doesn't stop at the first match, maybe that's intended so we can see all possible matches?

Reply:
Oh yeah you can completely ignore this code. As I wrote in the description this is just bonus for the showcase, but in no way meant to be merged. I just went for a quick implementation, because this needs to proper parsing logic and will change anyway.